hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/observer/ObservableBatch.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.impl.observer;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Objects;
+
+public class ObservableBatch implements IdentifiedDataSerializable {
+
+    private static final ObservableBatch END_OF_DATA = new ObservableBatch(null, null, true);
+    @Nullable
+    private Object[] items;
+    @Nullable
+    private Throwable throwable;
+    private boolean endOfData;
+
+    private ObservableBatch(@Nullable Object[] items, @Nullable Throwable throwable, boolean endOfData) {
+        this.items = items;
+        this.throwable = throwable;
+        this.endOfData = endOfData;
+    }
+
+    ObservableBatch() { //needed for deserialization
+    }
+
+    public static ObservableBatch items(@Nonnull ArrayList<Object> items) {
+        Objects.requireNonNull(items, "items");
+        return new ObservableBatch(items.toArray(), null, false);
+    }
+
+    public static ObservableBatch endOfData() {
+        return END_OF_DATA;
+    }
+
+    public static ObservableBatch error(Throwable throwable) {
+        return new ObservableBatch(null, Objects.requireNonNull(throwable, "throwable"), false);
+    }
+
+    @Nullable
+    Object[] getItems() {
+        return items;
+    }
+
+    @Nullable
+    Throwable getThrowable() {
+        return throwable;
+    }
+
+    boolean isEndOfData() {
+        return endOfData;
+    }
+
+    @Override
+    public int getFactoryId() {
+        return JetObserverDataSerializerHook.FACTORY_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return JetObserverDataSerializerHook.OBSERVABLE_BATCH;
+    }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
+        out.writeObject(items);
+        out.writeObject(throwable);
+        out.writeBoolean(endOfData);
```

Comment:
We can do a trick:
```java
out.writeBoolean(endOfData);
if (endOfData) {
    out.writeObject(throwable);
} else {
    out.writeObject(items);
}
```

And then in the read method:
```java
if (endOfData = in.readBoolean()) {
    items = in.readObject();
} else {
    throwable = in.readObject();
}
```
Note the single equation mark in the `if`, I'd rather extract it out for readability...

Reply:
I don't think we should. Almost always `throwable` will be null, which will end up executing `com.hazelcast.internal.serialization.impl.ConstantSerializers.NullSerializer`, which will write nothing, right? So we can just ignore `throwable` and then no tricks required. Do you agree?

Reply:
I'm not sure what it writes but it certainly isn't _nothing_. How would `readObject` then know that it's null? The readObject sees a stream of bytes and has to know how much to read.
That was my aim to save few bytes: at any time there's either items or throwable non-null, depending on the `endOfData` boolean. We should also assert that, if it's not true it will be hard to debug.